src/tool.cpp

```diff
@@ -87,6 +117,64 @@ void ToolCapabilities::deSerialize(std::istream &is)
 	}
 }
 
+void ToolCapabilities::serializeJson(std::ostream &os) const
+{
+	Json::Value root;
+	root["full_punch_interval"] = full_punch_interval;
+	root["max_drop_level"] = max_drop_level;
+
+	Json::Value groupcaps_object;
+	ToolGCMap::const_iterator gciter;
+	for (gciter = groupcaps.begin(); gciter != groupcaps.end(); ++gciter) {
+		gciter->second.toJson(groupcaps_object[gciter->first]);
+	}
+	root["groupcaps"] = groupcaps_object;
+
+	Json::Value damage_groups_object;
+	DamageGroup::const_iterator dgiter;
+	for (dgiter = damageGroups.begin(); dgiter != damageGroups.end(); ++dgiter) {
+		damage_groups_object[dgiter->first] = dgiter->second;
+	}
+	root["damage_groups"] = damage_groups_object;
+
+	os << root;
+}
+
+void ToolCapabilities::deserializeJson(std::istream &is)
+{
+	Json::Value root;
+	is >> root;
+	if (root.isObject()) {
```

Comment:
silently fails, should return false instead and then give a warning if the TOOLCAP_KEY string is non empty

Reply:
Do you mean give a warning where deserializeJson is used? (I don't remember where)

Reply:
Yeah, in updateToolCapabilities in itemstackmetadata. Or here is also fine actually, if you are able to get the length of IS (I forget if a stream could do this)

Reply:
If I give a warning here, should I still return false?

Reply:
If it's empty string, then wouldn't TOOLCAP_KEY not be contained?

Reply:
actually, nevermind - there's no need to show a warning as set_tool_capacities will always be used